src/common.jl

```diff
@@ -1,269 +1,250 @@
+        if isa(b, PSY.DCBranch)
+            @warn("PTDF construction ignores DC-Lines")
+            continue
+        end
+
+        (fr_b, to_b) = get_bus_indices(b, bus_lookup)
+        b_val = PSY.get_series_susceptance(b)
+
+        if fr_b ∉ slack_positions
+            check_ = sum(fr_b .> slack_positions)
+            push!(BA_I, ix)
+            push!(BA_J, fr_b - check_)
+            push!(BA_V, b_val)
+        end
+
+        if to_b ∉ slack_positions
+            check_ = sum(to_b .> slack_positions)
+            push!(BA_I, ix)
+            push!(BA_J, to_b - check_)
+            push!(BA_V, -b_val)
+        end
+    end
 
-# Handle the base case when we have fewer indices than lookups:
-_to_index_tuple(idx::NTuple{0}, lookup::Tuple) = ()
-
-# Resolve ambiguity with the above two base cases
-_to_index_tuple(idx::NTuple{0}, lookup::NTuple{0}) = ()
-to_index(A::PowerNetworkMatrix, idx...) = _to_index_tuple(idx, A.lookup)
-
-# Doing `Colon() in idx` is relatively slow because it involves
-# a non-unrolled loop through the `idx` tuple which may be of
-# varying element type. Another lisp-y recursion trick fixes that
-has_colon(idx::Tuple{}) = false
-has_colon(idx::Tuple) = isa(first(idx), Colon) || has_colon(Base.tail(idx))
-
-# TODO: better error (or just handle correctly) when user tries to index with a range like a:b
-# The only kind of slicing we support is dropping a dimension with colons
-function Base.getindex(A::PowerNetworkMatrix, row, column)
-    #=
-    This is old code when we accepted idx... instead of row, column
-    if has_colon(idx)
-        PTDF(A.data[to_index(A,idx...)...], (ax for (i,ax) in enumerate(A.axes) if idx[i] == Colon())...)
+    BA = SparseArrays.sparse(BA_I, BA_J, BA_V)
+
+    return BA
+end
+
+# ABA matrix evaluation ######################################################
+function calculate_ABA_matrix(
+    A::SparseArrays.SparseMatrixCSC{Int8, Int32},
+    BA::SparseArrays.SparseMatrixCSC{T, Int32} where {T <: Union{Float32, Float64}},
+    slack_positions::Vector{Int64})
+    return A[:, setdiff(1:end, slack_positions[1])]' * BA
+end
+
+# PTDF evaluation ############################################################
+function calculate_PTDF_matrix_KLU(branches, nodes::Vector{PSY.Bus},
+    dist_slack::Vector{Float64})
+    buscount = length(nodes)
+    linecount = length(branches)
+
+    A, slack_positions = calculate_A_matrix(branches, nodes)
+    bus_lookup = _make_ax_ref(nodes)
+    BA = calculate_BA_matrix(branches, slack_positions, bus_lookup)
+    ABA = calculate_ABA_matrix(A, BA, slack_positions)
+    K = klu(ABA)
+    Ix = Matrix(1.0LinearAlgebra.I, size(ABA, 1), size(ABA, 1))
+    ABA_inv = zeros(Float64, size(ABA, 1), size(ABA, 2))
+    ldiv!(ABA_inv, K, Ix)
+
+    slack_position = slack_positions[1]
+    if dist_slack[1] == 0.1 && length(dist_slack) == 1
+        S = BA * ABA_inv
+        @views S =
+            hcat(S[:, 1:(slack_position - 1)], zeros(linecount), S[:, slack_position:end])
```

Comment:
the hcat operation here is allocating a new matrix

Reply:
solved